File: bindings/dotnet/DotOpenDAL.Tests/LayerTest.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

using DotOpenDAL.Layer;

namespace DotOpenDAL.Tests;

public class LayerTest
{
    [Fact]
    public void WithConcurrentLimit_ReturnsNewOperator()
    {
        using var op = new Operator("memory");
        var before = op.Op;
        using var layered = op.WithLayer(new ConcurrentLimitLayer(4));

        Assert.NotEqual(IntPtr.Zero, layered.Op);
        Assert.NotSame(op, layered);
        Assert.Equal(before, op.Op);
        Assert.NotEqual(before, layered.Op);

        layered.Write("layer-concurrent", [1, 2, 3]);
        var value = layered.Read("layer-concurrent");
        Assert.Equal([1, 2, 3], value);
    }

    [Fact]
    public void WithRetry_ReturnsNewOperator()
    {
        using var op = new Operator("memory");
        var before = op.Op;
        using var layered = op.WithLayer(new RetryLayer
        {
            Jitter = false,
            Factor = 2,
            MinDelay = TimeSpan.FromMilliseconds(1),
            MaxDelay = TimeSpan.FromMilliseconds(10),
            MaxTimes = 2,
        });

        Assert.NotEqual(IntPtr.Zero, layered.Op);
        Assert.NotSame(op, layered);
        Assert.Equal(before, op.Op);
        Assert.NotEqual(before, layered.Op);

        layered.Write("layer-retry", [4, 5, 6]);
        var value = layered.Read("layer-retry");
        Assert.Equal([4, 5, 6], value);
    }

    [Fact]
    public void WithConcurrentLimit_ZeroPermits_ThrowsArgumentOutOfRangeException()
    {
        using var op = new Operator("memory");

        Assert.Throws<ArgumentOutOfRangeException>(() => op.WithLayer(new ConcurrentLimitLayer(0)));
    }

    [Fact]
    public void WithRetry_InvalidFactor_ThrowsArgumentOutOfRangeException()
    {
        using var op = new Operator("memory");

        Assert.Throws<ArgumentOutOfRangeException>(() => op.WithLayer(new RetryLayer
        {
            Factor = 0,
        }));
    }

    [Fact]
    public void WithTimeout_ReturnsNewOperator()
    {
        using var op = new Operator("memory");
        var before = op.Op;
        using var layered = op.WithLayer(new TimeoutLayer
        {
            Timeout = TimeSpan.FromSeconds(5),
            IoTimeout = TimeSpan.FromSeconds(2),
        });

        Assert.NotEqual(IntPtr.Zero, layered.Op);
        Assert.NotSame(op, layered);
        Assert.Equal(before, op.Op);
        Assert.NotEqual(before, layered.Op);

        layered.Write("layer-timeout", [7, 8, 9]);
        var value = layered.Read("layer-timeout");
        Assert.Equal([7, 8, 9], value);
    }

    [Fact]
    public void WithTimeout_ZeroTimeout_ThrowsArgumentOutOfRangeException()
    {
        using var op = new Operator("memory");

        Assert.Throws<ArgumentOutOfRangeException>(() => op.WithLayer(new TimeoutLayer
        {
            Timeout = TimeSpan.Zero,
        }));
    }

    [Fact]
    public void WithLayer_OperatorsCanBeDisposedIndependently()
    {
        var op = new Operator("memory");
        var layered = op.WithLayer(new ConcurrentLimitLayer(2));

        layered.Dispose();

        op.Write("layer-dispose-origin", [1, 1, 1]);
        var originalValue = op.Read("layer-dispose-origin");
        Assert.Equal([1, 1, 1], originalValue);

        op.Dispose();

        var op2 = new Operator("memory");
        var layered2 = op2.WithLayer(new ConcurrentLimitLayer(2));

        op2.Dispose();

        layered2.Write("layer-dispose-layered", [2, 2, 2]);
        var layeredValue = layered2.Read("layer-dispose-layered");
        Assert.Equal([2, 2, 2], layeredValue);

        layered2.Dispose();
    }
}

File: bindings/cpp/tests/layer_test.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include <chrono>
#include <memory>
#include <utility>
#include <vector>

#include "framework/test_framework.hpp"
#include "layer.hpp"
#include "mocks/mock_layer_builder.hpp"

namespace opendal::test {

class LayerTest : public ::testing::Test {};

OPENDAL_TEST_F(LayerTest, WithTimeoutForwardsNanosecondsToMutator) {
  MockLayerBuilderMutator mutator;
  const auto timeout = std::chrono::seconds(30);
  const auto io_timeout = std::chrono::seconds(5);

  EXPECT_CALL(
      mutator,
      AddTimeout(static_cast<uint64_t>(std::chrono::duration_cast<std::chrono::nanoseconds>(
                                           timeout)
                                           .count()),
                 static_cast<uint64_t>(std::chrono::duration_cast<std::chrono::nanoseconds>(
                                             io_timeout)
                                             .count())))
      .Times(1);

  WithTimeout(timeout, io_timeout)->ApplyTo(mutator);
}

OPENDAL_TEST_F(LayerTest, WithRetryForwardsConfigToMutator) {
  MockLayerBuilderMutator mutator;

  EXPECT_CALL(mutator, AddRetry(false, 2.0F, 1'000'000ULL, 10'000'000ULL, 5))
      .Times(1);

  RetryConfig config;
  config.max_times = 5;
  config.min_delay = std::chrono::milliseconds(1);
  config.max_delay = std::chrono::milliseconds(10);

  WithRetry(config)->ApplyTo(mutator);
}

OPENDAL_TEST_F(LayerTest, WithRetryEnablesJitter) {
  MockLayerBuilderMutator mutator;

  EXPECT_CALL(mutator, AddRetry(true, testing::_, testing::_, testing::_, testing::_))
      .Times(1);

  RetryConfig config;
  config.jitter = true;
  WithRetry(config)->ApplyTo(mutator);
}

OPENDAL_TEST_F(LayerTest, WithRetryUsesDefaultConfig) {
  MockLayerBuilderMutator mutator;

  EXPECT_CALL(mutator,
              AddRetry(false, 2.0F, 1'000'000'000ULL, 60'000'000'000ULL, 3))
      .Times(1);

  WithRetry()->ApplyTo(mutator);
}

OPENDAL_TEST_F(LayerTest, LayerOptionsApplyInOrder) {
  MockLayerBuilderMutator mutator;
  testing::InSequence sequence;

  EXPECT_CALL(mutator, AddTimeout(testing::_, testing::_)).Times(1);
  EXPECT_CALL(mutator, AddRetry(testing::_, testing::_, testing::_, testing::_, testing::_))
      .Times(1);

  WithTimeout(std::chrono::minutes(1), std::chrono::seconds(10))->ApplyTo(mutator);
  WithRetry()->ApplyTo(mutator);
}

OPENDAL_TEST_F(LayerTest, OperatorWithTimeoutAndRetry) {
  std::vector<std::unique_ptr<OperatorOption>> options;
  options.push_back(WithTimeout(std::chrono::minutes(1),
                                std::chrono::seconds(10)));
  options.push_back(WithRetry());

  Operator op("memory", {}, std::move(options));
  ASSERT_TRUE(op.Available());

  constexpr std::string_view data = "layer-test";
  op.Write("layer-test.txt", data);
  EXPECT_EQ(op.Read("layer-test.txt"), data);
}

OPENDAL_TEST_F(LayerTest, WithTimeoutRejectsZeroTimeout) {
  std::vector<std::unique_ptr<OperatorOption>> options;
  options.push_back(WithTimeout(std::chrono::seconds(0), std::chrono::seconds(1)));
  EXPECT_THROW(Operator("memory", {}, std::move(options)), std::invalid_argument);
}

OPENDAL_TEST_F(LayerTest, WithTimeoutRejectsZeroIoTimeout) {
  std::vector<std::unique_ptr<OperatorOption>> options;
  options.push_back(WithTimeout(std::chrono::seconds(1), std::chrono::seconds(0)));
  EXPECT_THROW(Operator("memory", {}, std::move(options)), std::invalid_argument);
}

OPENDAL_TEST_F(LayerTest, RetryMaxTimesRejectsZero) {
  RetryConfig config;
  config.max_times = 0;

  std::vector<std::unique_ptr<OperatorOption>> options;
  options.push_back(WithRetry(config));

  EXPECT_THROW(Operator("memory", {}, std::move(options)), std::invalid_argument);
}

OPENDAL_TEST_F(LayerTest, RetryMaxDelayBeforeMinDelay) {
  RetryConfig config;
  config.min_delay = std::chrono::seconds(10);
  config.max_delay = std::chrono::seconds(1);

  std::vector<std::unique_ptr<OperatorOption>> options;
  options.push_back(WithRetry(config));

  EXPECT_THROW(Operator("memory", {}, std::move(options)), std::invalid_argument);
}

OPENDAL_TEST_F(LayerTest, RetryFactorRejectsInvalidValue) {
  RetryConfig config;
  config.factor = 0.5F;

  std::vector<std::unique_ptr<OperatorOption>> options;
  options.push_back(WithRetry(config));

  EXPECT_THROW(Operator("memory", {}, std::move(options)), std::invalid_argument);
}

}  // namespace opendal::test
